### PR TITLE
feat: Add transform dereference result hook

### PR DESCRIPTION
**Pull request by pytlesk4**

### CODE REVIEW

File: src/runner.ts [hunk 1]
```diff
@@ -408,6 +412,22 @@ export class ResolveRunner implements Types.IResolveRunner {
...
+          const { result, error } = await this.transformDereferenceResult({
+            source: uriResolver.source,
+            result: lookupResult.resolved.result,
+            targetAuthority: ref,
+            parentAuthority: this.baseUri,
+            parentPath,
+            fragment: ref.fragment(),
+          });
+          if (error) {
+            throw new Error(`Could not transform dereferenced result for '${ref.toString()}' - ${String(error)}`);
```
Comment (marbemac) [inline src/runner.ts]: Perhaps could set `lookupResult.error` rather than throw?

File: src/types.ts [hunk 2]
```diff
@@ -115,7 +122,7 @@ export interface IResolver {
 export interface IUriParser {
   result: any;
   fragment: string;
-  uriResult: IUriResult;
+  // uriResult: IUriResult;
```
Comment (marbemac) [inline src/types.ts]: Why?

File: src/runner.ts [hunk 3]
```diff
@@ -408,6 +412,22 @@ export class ResolveRunner implements Types.IResolveRunner {
       if (uriResolver) {
         lookupResult.resolved = await uriResolver.resolve(Utils.uriToJSONPointer(ref));
 
+        // support custom transformers
+        if (this.transformDereferenceResult) {
```
Comment (marbemac) [inline src/runner.ts]: Probably better in the resolve function? https://github.com/stoplightio/json-ref-resolver/blob/master/src/runner.ts#L235

File: src/runner.ts [hunk 4]
```diff
@@ -38,6 +38,7 @@ export class ResolveRunner implements Types.IResolveRunner {
   public readonly getRef: (key: string, val: any) => string | void;
   public readonly transformRef?: (opts: Types.IRefTransformer, ctx: any) => uri.URI | any;
   public readonly parseResolveResult?: (opts: Types.IUriParser) => Promise<Types.IUriParserResult>;
+  public readonly transformDereferenceResult?: (opts: Types.ITransformerOptions) => Promise<Types.ITransformerResult>;
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply src/runner.ts]: done

File: src/runner.ts [hunk 5]
```diff
@@ -408,6 +412,22 @@ export class ResolveRunner implements Types.IResolveRunner {
...
+        if (this.transformDereferenceResult) {
+          const { result, error } = await this.transformDereferenceResult({
+            source: uriResolver.source,
+            result: lookupResult.resolved.result,
+            targetAuthority: ref,
+            parentAuthority: this.baseUri,
+            parentPath,
+            fragment: ref.fragment(),
+          });
+          if (error) {
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply src/runner.ts]: Done

File: src/runner.ts (hunk 1 above)
Comment (pytlesk4) [reply src/runner.ts]: Let me know what you think now after moving where this get's called at.

File: src/runner.ts (hunk 3 above)
Comment (pytlesk4) [reply src/runner.ts]: Done

File: src/runner.ts [hunk 6]
```diff
@@ -318,7 +320,8 @@ export class ResolveRunner implements Types.IResolveRunner {
     if (this.parseResolveResult) {
       try {
         const parsed = await this.parseResolveResult({
-          uriResult: result,
+          // TODO: Is this correct? Result has type any, but uriResult should be type IUriResult.
+          // uriResult: result,
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply src/runner.ts]: `uriResult` here is a just an object, but according to types.ts, it should be `IUriResult`.

result comes from this [line](https://github.com/stoplightio/json-ref-resolver/blob/f18cebe445717c602a5055f81c8ecf5dd2e703d1/src/runner.ts#L347)

[the remark this answers is not in the document]
Comment (pytlesk4) [reply src/runner.ts]: I undid my changes, but we should at the very least update types.ts probably.

File: src/types.ts (hunk 2 above)
Comment (pytlesk4) [reply src/types.ts]: Answered above

**Pull request merged**